### PR TITLE
Fix various reference cycles in the examples

### PR DESCRIPTION
**Pull request by sdroege**
Fixes https://github.com/gtk-rs/examples/issues/179

### CODE REVIEW

File: src/bin/grid.rs [hunk 1]
```diff
@@ -15,6 +15,19 @@ mod example {
 
     use std::env::args;
 
+    // upgrade weak reference or return
+    #[macro_export]
+    macro_rules! upgrade_weak {
```
Comment (GuillaumeGomez) [inline src/bin/grid.rs]: 👍

**Comment by GuillaumeGomez** [comment]
Thanks! At some point, we'll also need to explain where people should `clone` and where they should `downgrade`/`upgrade`.

File: src/bin/listbox_model.rs [hunk 2]
```diff
@@ -140,7 +159,9 @@ fn build_ui(application: &gtk::Application) {
 
         // When a row is activated (select + enter) we simply emit the clicked
         // signal on the corresponding edit button to open the edit dialog
-        box_.connect_activate(clone!(edit_button => move |_| {
+        let edit_button_weak = edit_button.downgrade();
+        box_.connect_activate(clone!(edit_button_weak => move |_| {
```
Comment (EPashkin) [inline src/bin/listbox_model.rs]: There was really need use `clone!` ?

File: src/bin/grid.rs (hunk 1 above)
Comment (EPashkin) [reply src/bin/grid.rs]: Seems this enough specific macro to be moved to glib

File: src/bin/listbox_model.rs (hunk 2 above)
Comment (sdroege) [reply src/bin/listbox_model.rs]: No, actually not

File: src/bin/grid.rs (hunk 1 above)
Comment (sdroege) [reply src/bin/grid.rs]: It also works on `Arc` and `Rc` btw

Comment (GuillaumeGomez) [reply src/bin/grid.rs]: Indeed, I'll let one of you add it to glib. ;) (and update examples after hehe)

**Pull request merged**